File: ViewModel/GridControlViewModel.cs
using System;
using System.ComponentModel;
using System.Runtime.CompilerServices;
using System.Windows;
using System.Windows.Input;
using Crossword.Services;

namespace Crossword.ViewModel;

public class GridControlViewModel : ViewModelBase
{
    private readonly IDialogService _dialogService;
    private readonly IGridManagerService _gridManagerService;
    private int _numberOfCellsHorizontally = 30;
    private int _numberOfCellsVertically = 30;
    private bool _isVerticallyMirror;
    private bool _isHorizontallyMirror;
    private bool _isAllMirror;
    private bool _isVerticallyMirrorRevers;
    private bool _isHorizontallyMirrorRevers;
    public ICommand ChangeFieldSizeCommand { get; }
    public ICommand CellInteractionCommand { get; }

    public GridControlViewModel(IDialogService dialogService, IGridManagerService gridManagerService, Func<bool> canInteract)
    {
        _dialogService = dialogService;
        _gridManagerService = gridManagerService;

        ChangeFieldSizeCommand = new RelayCommand(_ => ChangeFieldSize());
        CellInteractionCommand = new RelayCommand(HandleCellInteraction, _ => canInteract());

        _gridManagerService.PropertyChanged += OnGridManagerPropertyChanged;
        UpdateMirrorLines();
    }

    private void OnGridManagerPropertyChanged(object? sender, PropertyChangedEventArgs e)
    {
        if (e.PropertyName == nameof(IGridManagerService.FieldWidth)) OnPropertyChanged(nameof(FieldWidth));
        if (e.PropertyName == nameof(IGridManagerService.FieldHeight)) OnPropertyChanged(nameof(FieldHeight));
    }

    public int FieldWidth => _gridManagerService.FieldWidth;
    public int FieldHeight => _gridManagerService.FieldHeight;

    #region Mirror Properties

    public bool IsVerticallyMirror
    {
        get => _isVerticallyMirror;
        set => SetProperty(ref _isVerticallyMirror, value);
    }

    public bool IsHorizontallyMirror
    {
        get => _isHorizontallyMirror;
        set => SetProperty(ref _isHorizontallyMirror, value);
    }

    public bool IsAllMirror
    {
        get => _isAllMirror;
        set => SetProperty(ref _isAllMirror, value);
    }

    public bool IsVerticallyMirrorRevers
    {
        get => _isVerticallyMirrorRevers;
        set => SetProperty(ref _isVerticallyMirrorRevers, value);
    }

    public bool IsHorizontallyMirrorRevers
    {
        get => _isHorizontallyMirrorRevers;
        set => SetProperty(ref _isHorizontallyMirrorRevers, value);
    }

    public Visibility LineCenterVVisibility => (IsVerticallyMirror || IsVerticallyMirrorRevers || IsAllMirror) ? Visibility.Visible : Visibility.Hidden;
    public Visibility LineCenterHVisibility => (IsHorizontallyMirror || IsHorizontallyMirrorRevers || IsAllMirror) ? Visibility.Visible : Visibility.Hidden;
    public double LineCenterH_X => (_numberOfCellsHorizontally * CellViewModel.CellSize / 2.0) + CellViewModel.CellSize;
    public double LineCenterH_Y2 => (_numberOfCellsVertically + 1) * CellViewModel.CellSize;
    public double LineCenterV_Y => (_numberOfCellsVertically * CellViewModel.CellSize / 2.0) + CellViewModel.CellSize;
    public double LineCenterV_X2 => (_numberOfCellsHorizontally + 1) * CellViewModel.CellSize;

    #endregion

    private void HandleCellInteraction(object? param)
    {
        if (param is not CellViewModel cellVm) return;

        _gridManagerService.SetMirrorModes(IsVerticallyMirror, IsHorizontallyMirror, IsAllMirror, IsVerticallyMirrorRevers, IsHorizontallyMirrorRevers);
        _gridManagerService.HandleCellInteraction(cellVm);
    }

    private void ChangeFieldSize()
    {
        _dialogService.ShowChangeFillDialog(ref _numberOfCellsHorizontally, ref _numberOfCellsVertically);
        _gridManagerService.InitializeGrid(_numberOfCellsHorizontally, _numberOfCellsVertically);
        UpdateMirrorLines();
    }

    private void UpdateMirrorLines()
    {
        OnPropertyChanged(nameof(LineCenterVVisibility));
        OnPropertyChanged(nameof(LineCenterHVisibility));
        OnPropertyChanged(nameof(LineCenterH_X));
        OnPropertyChanged(nameof(LineCenterH_Y2));
        OnPropertyChanged(nameof(LineCenterV_Y));
        OnPropertyChanged(nameof(LineCenterV_X2));
    }

    protected override void OnPropertyChanged([CallerMemberName] string? propertyName = null)
    {
        base.OnPropertyChanged(propertyName);
        if (propertyName?.Contains("Mirror") ?? false)
        {
            UpdateMirrorLines();
        }
    }
}

File: crossword_python/viewmodels/GridControlViewModel.py
from crossword_python.core.viewmodel_base import ViewModelBase
from crossword_python.core.command import RelayCommand
from crossword_python.viewmodels.CellViewModel import CellViewModel

class GridControlViewModel(ViewModelBase):
    def __init__(self, dialog_service, grid_manager_service, can_interact_func):
        super().__init__()
        self._dialog_service = dialog_service
        self._grid_manager_service = grid_manager_service
        self._can_interact_func = can_interact_func
        
        self._number_of_cells_horizontally = 30
        self._number_of_cells_vertically = 30

        self._is_clear_mirror = True
        self._is_vertically_mirror = False
        self._is_horizontally_mirror = False
        self._is_all_mirror = False
        self._is_vertically_mirror_revers = False
        self._is_horizontally_mirror_revers = False

        self.change_field_size_command = RelayCommand(self.change_field_size)
        self.cell_interaction_command = RelayCommand(self.handle_cell_interaction, lambda _: self._can_interact_func())
        self.clear_grid_command = RelayCommand(self.clear_grid, lambda _: self._can_interact_func())

        self._grid_manager_service.property_changed.connect(self.on_grid_manager_property_changed)
        self.update_mirror_lines()

    def on_grid_manager_property_changed(self, prop_name: str):
        if prop_name in ["field_width", "field_height"]:
            self.property_changed.emit(prop_name)

    @property
    def field_width(self) -> int:
        return self._grid_manager_service.field_width

    @property
    def field_height(self) -> int:
        return self._grid_manager_service.field_height

    # Зеркальные свойства
    @property
    def is_clear_mirror(self) -> bool: return self._is_clear_mirror
    @is_clear_mirror.setter
    def is_clear_mirror(self, value: bool):
        if self.set_property('_is_clear_mirror', value): self.update_mirror_lines()

    @property
    def is_vertically_mirror(self) -> bool: return self._is_vertically_mirror
    @is_vertically_mirror.setter
    def is_vertically_mirror(self, value: bool):
        if self.set_property('_is_vertically_mirror', value): self.update_mirror_lines()

    @property
    def is_horizontally_mirror(self) -> bool: return self._is_horizontally_mirror
    @is_horizontally_mirror.setter
    def is_horizontally_mirror(self, value: bool):
        if self.set_property('_is_horizontally_mirror', value): self.update_mirror_lines()

    @property
    def is_all_mirror(self) -> bool: return self._is_all_mirror
    @is_all_mirror.setter
    def is_all_mirror(self, value: bool):
        if self.set_property('_is_all_mirror', value): self.update_mirror_lines()

    @property
    def is_vertically_mirror_revers(self) -> bool: return self._is_vertically_mirror_revers
    @is_vertically_mirror_revers.setter
    def is_vertically_mirror_revers(self, value: bool):
        if self.set_property('_is_vertically_mirror_revers', value): self.update_mirror_lines()

    @property
    def is_horizontally_mirror_revers(self) -> bool: return self._is_horizontally_mirror_revers
    @is_horizontally_mirror_revers.setter
    def is_horizontally_mirror_revers(self, value: bool):
        if self.set_property('_is_horizontally_mirror_revers', value): self.update_mirror_lines()

    # Свойства линий центра
    @property
    def line_center_v_visible(self) -> bool:
        return self.is_vertically_mirror or self.is_vertically_mirror_revers or self.is_all_mirror

    @property
    def line_center_h_visible(self) -> bool:
        return self.is_horizontally_mirror or self.is_horizontally_mirror_revers or self.is_all_mirror

    @property
    def line_center_h_x(self) -> float:
        return (self._number_of_cells_horizontally * CellViewModel.CELL_SIZE / 2.0) + CellViewModel.CELL_SIZE

    @property
    def line_center_h_y2(self) -> float:
        return (self._number_of_cells_vertically + 1) * CellViewModel.CELL_SIZE

    @property
    def line_center_v_y(self) -> float:
        return (self._number_of_cells_vertically * CellViewModel.CELL_SIZE / 2.0) + CellViewModel.CELL_SIZE

    @property
    def line_center_v_x2(self) -> float:
        return (self._number_of_cells_horizontally + 1) * CellViewModel.CELL_SIZE

    def handle_cell_interaction(self, params: tuple):
        cell_vm, new_color = params
        self._grid_manager_service.set_mirror_modes(
            self.is_vertically_mirror, self.is_horizontally_mirror, 
            self.is_all_mirror, self.is_vertically_mirror_revers, self.is_horizontally_mirror_revers
        )
        self._grid_manager_service.handle_cell_interaction(cell_vm, new_color)

    def change_field_size(self, parameter=None):
        result, new_h, new_v = self._dialog_service.show_change_fill_dialog(
            self._number_of_cells_horizontally, self._number_of_cells_vertically
        )
        if result:
            self._number_of_cells_horizontally = new_h
            self._number_of_cells_vertically = new_v
            self._grid_manager_service.initialize_grid(self._number_of_cells_horizontally, self._number_of_cells_vertically)
            self.update_mirror_lines()

    def clear_grid(self, parameter=None):
        self._grid_manager_service.initialize_grid(self._number_of_cells_horizontally, self._number_of_cells_vertically)
        self.update_mirror_lines()

    def update_mirror_lines(self):
        self.property_changed.emit("line_center_v_visible")
        self.property_changed.emit("line_center_h_visible")
        self.property_changed.emit("line_center_h_x")
        self.property_changed.emit("line_center_h_y2")
        self.property_changed.emit("line_center_v_y")
        self.property_changed.emit("line_center_v_x2")